src/bootstrap/compile.rs

```diff
@@ -248,11 +245,7 @@ impl Step for StdLink {
             copy_apple_sanitizer_dylibs(builder, &builder.native_dir(target), "osx", &libdir);
         }
 
-        builder.ensure(tool::CleanTools {
-            compiler: target_compiler,
-            target,
-            cause: Mode::Std,
-        });
+        builder.cargo(target_compiler, Mode::ToolStd, target, "clean");
```

Comment:
Hm, is this (and similar statements below) actually necessary? I would've hoped this would happen on its own.

Reply:
Step resolution shouldn't depend on files existing on the system - do you happen to have a concrete example? If we actually do need this then can we factor out the cleaning logic from `Builder::cargo` and call that separately here? It feels a little odd to be creating the full command and then not running it (i.e., using implicit side-effects).

Reply:
If you don't feel like dealing with this though feel free to let me know -- I'm fine with leaving these in for now

Reply:
Yeah, `i.e` removing both https://github.com/rust-lang/rust/pull/52036/files#diff-fa9668c926a2788f7849514fe3ed66a9R248 for `StdLink` and https://github.com/rust-lang/rust/pull/52036/files#diff-fa9668c926a2788f7849514fe3ed66a9R446 for `TestLink` then doing `./x.py test src/bootstrap`, Assertion https://github.com/rust-lang/rust/blob/1ecf6929dc3b309cdfcb7239260777dab38242b9/src/bootstrap/builder.rs#L1660 will fail and looking at diff shows one of steps was missed. Perhaps we could factor out cleaning logic as you had suggested.

Reply:
Okay, let's leave this as is for now and look into the details later -- it seems fine to leave this somewhat half done pending a more thorough investigation as to how *exactly* the dependency edge is created.